### PR TITLE
WT-11413 Add histogram stats for compression ratio of blocks written to disk

### PR DESCRIPTION
**Pull request by ajmorton**
Two changes in this PR:
1. Add new histogram stats that report block compression ratios when we write blocks to disk
2. Reword some stat names and description for clarity. The existing `compress_hist_ratio_*` stats are now called `compress_read_ratio_hist_*` and compression stats now talk about compressed blocks rather than compressed pages

I'll create a separate ticket for testing as per the discussion on the Jira.

For an example of the new stats, here is an extract from the WiredTigerStat file during a test/format run
```
"compression":{
    "compressed page maximum internal page size prior to compression":131072,
    "compressed page maximum leaf page size prior to compression ":131072,
    "page written to disk failed to compress":0,
    "page written to disk was too small to compress":0,
    "pages read from disk":0,
    "pages read from disk with compression ratio greater than 64":0,
    "pages read from disk with compression ratio smaller than  2":0,
    "pages read from disk with compression ratio smaller than  4":0,
    "pages read from disk with compression ratio smaller than  8":0,
    "pages read from disk with compression ratio smaller than 16":0,
    "pages read from disk with compression ratio smaller than 32":0,
    "pages read from disk with compression ratio smaller than 64":0,
    "pages written to disk":0,
    "pages written to disk with compression ratio greater than 64":0,
    "pages written to disk with compression ratio smaller than  2":0,
    "pages written to disk with compression ratio smaller than  4":0,
    "pages written to disk with compression ratio smaller than  8":0,
    "pages written to disk with compression ratio smaller than 16":0,
    "pages written to disk with compression ratio smaller than 32":0,
    "pages written to disk with compression ratio smaller than 64":0
}
```

### CODE REVIEW

File: dist/stat_data.py [hunk 1]
```diff
@@ -820,19 +820,26 @@ def __init__(self, name, desc, flags=''):
...
-    CompressStat('compress_hist_ratio_64', 'number of blocks with compress ratio smaller than 64'),
-    CompressStat('compress_hist_ratio_max', 'number of blocks with compress ratio greater than 64'),
     CompressStat('compress_precomp_intl_max_page_size', 'compressed page maximum internal page size prior to compression', 'no_clear,no_scale,size'),
     CompressStat('compress_precomp_leaf_max_page_size', 'compressed page maximum leaf page size prior to compression ', 'no_clear,no_scale,size'),
-    CompressStat('compress_read', 'compressed pages read'),
-    CompressStat('compress_write', 'compressed pages written'),
-    CompressStat('compress_write_fail', 'page written failed to compress'),
-    CompressStat('compress_write_too_small', 'page written was too small to compress'),
+    CompressStat('compress_read', 'compressed blocks read'),
+    CompressStat('compress_read_ratio_hist_2', 'compressed blocks read with compression ratio smaller than  2'),
```
Comment (ajmorton) [inline dist/stat_data.py:826]: Small stylistic note. Since stats or sorted alphabetically on their descriptions I've slipped in two spaces before the single digits ranges (2,4,8). This means the stats are sorted 2,4,8,16,32,64 instead of 16,2,32,4,64,8

File: dist/stat_data.py [hunk 2]
```diff
@@ -820,19 +820,26 @@ def __init__(self, name, desc, flags=''):
...
+    CompressStat('compress_read_ratio_hist_max', 'compressed blocks read with compression ratio greater than 64'),
+    CompressStat('compress_write', 'compressed blocks written'),
+    CompressStat('compress_write_ratio_hist_2', 'compressed blocks written with compression ratio smaller than  2'),
+    CompressStat('compress_write_ratio_hist_4', 'compressed blocks written with compression ratio smaller than  4'),
+    CompressStat('compress_write_ratio_hist_8', 'compressed blocks written with compression ratio smaller than  8'),
+    CompressStat('compress_write_ratio_hist_16', 'compressed blocks written with compression ratio smaller than 16'),
+    CompressStat('compress_write_ratio_hist_32', 'compressed blocks written with compression ratio smaller than 32'),
+    CompressStat('compress_write_ratio_hist_64', 'compressed blocks written with compression ratio smaller than 64'),
+    CompressStat('compress_write_ratio_hist_max', 'compressed blocks written with compression ratio greater than 64'),
+    CompressStat('compress_write_fail', 'block written failed to compress'),
```
Comment (ajmorton) [inline dist/stat_data.py:841]: Another thing to call out is in the descriptions I've changed references from "pages compressed" to "blocks compressed" since the surrounding code refers to blocks rather than pages. This makes sense for WT devs but I'm interested on input if this is helpful/harmful for other T2 users to see blocks instead of pages

File: dist/stat_data.py (hunk 1 above)
Comment (etienneptl) [reply dist/stat_data.py:826]: I wonder if this deserves a comment in the file so no one treats it as a mistake.

File: dist/stat_data.py (hunk 2 above)
Comment (etienneptl) [reply dist/stat_data.py:841]: @agorrod, it seems that we are using pages and blocks interchangeably while I agree that blocks should be used here. What do you think?

File: src/include/stat.h [hunk 3]
```diff
@@ -1138,19 +1159,26 @@ struct __wt_dsrc_stats {
...
+    int64_t compress_read_ratio_hist_64;
+    int64_t compress_write;
+    int64_t compress_write_ratio_hist_max;
+    int64_t compress_write_ratio_hist_2;
+    int64_t compress_write_ratio_hist_4;
+    int64_t compress_write_ratio_hist_8;
+    int64_t compress_write_ratio_hist_16;
+    int64_t compress_write_ratio_hist_32;
+    int64_t compress_write_ratio_hist_64;
+    int64_t compress_precomp_intl_max_page_size;
```
Comment (etienneptl) [inline src/include/stat.h:1180]: The A->Z sorting seems wrong here

**Review by etienneptl — APPROVED** [review]
LGTM % 1 comment. @agorrod, I am requesting your review here as you were involved in the first PR I started but never finished.

File: src/include/stat.h (hunk 3 above)
Comment (ajmorton) [reply src/include/stat.h:1180]: Yep that's because this is auto-generated by `dist/stat.py` and the script sorts on [description text instead of variable name](https://github.com/wiredtiger/wiredtiger/blob/b311ca93ee650984e9104a250efb3e7bf431ac6c/dist/stat.py#L38). It was one of the motivations for rewording descriptions to give us nicer ordering of variables.

File: dist/stat_data.py (hunk 1 above)
Comment (ajmorton) [reply dist/stat_data.py:829]: SGTM 👍. Updated in [e544af4](https://github.com/wiredtiger/wiredtiger/pull/9908/commits/e544af4b96440d7ff122e6a40fb7fcf85334bf0c)

File: dist/stat_data.py (hunk 2 above)
Comment (agorrod) [reply dist/stat_data.py:844]: My preference is to stick with "page" generally (though I probably am guilty of being inconsistent with that).

Technically: the block manager allocates and tracks space utilization in the file. That space is tracked as blocks - where it is agnostic to the content being stored (hence it's called a block, instead of a page - which has a known structure and content).

There isn't a clear line in this case, but I think of compression as at the page level. i.e: we've laid out a page and are now compressing that content before storing it at the right place in the file via the block manager. On the other hand we do term this "block compression" in some places - hence personal preference comes in.

tl;dr:
I have a strong preference for consistency (where it doesn't involve wholesale changes of things already in the field), I have a mild preference for "page", but could be convinced that "block" is the better term here. If we go with page, maybe an additional "on-disk page" is worthwhile here, since we have so many different types of pages..

Comment (ajmorton) [reply dist/stat_data.py]: SGTM. Most of our existing statistics - other than the compression histogram stats that we want to reword - use pages already so I'll align with existing usage. I've changed back to using pages in [5cf36f6](https://github.com/wiredtiger/wiredtiger/pull/9908/commits/5cf36f69289817a2b05ed4c43b994cc1fc16ab1b)

File: dist/stat_data.py [hunk 4]
```diff
@@ -825,19 +825,29 @@ def __init__(self, name, desc, flags=''):
...
     CompressStat('compress_precomp_leaf_max_page_size', 'compressed page maximum leaf page size prior to compression ', 'no_clear,no_scale,size'),
-    CompressStat('compress_read', 'compressed pages read'),
-    CompressStat('compress_write', 'compressed pages written'),
-    CompressStat('compress_write_fail', 'page written failed to compress'),
-    CompressStat('compress_write_too_small', 'page written was too small to compress'),
+    CompressStat('compress_read', 'compressed pages read from disk'),
+    # dist/stat.py sorts stats by their descriptions and not their names. The following stat descriptions insert an extra 
+    # space before the single digit numbers (2, 4, 8) so stats will be sorted numerically (2, 4, 8, 16, 32) instead of 
+    # alphabetically (16, 2, 32, 4, 8).
+    CompressStat('compress_read_ratio_hist_2', 'compressed pages read from disk with compression ratio smaller than  2'),
```
Comment (agorrod) [inline dist/stat_data.py:834]: I don't remember, but does the "CompressStat" already add a standard prefix? i.e: Will the word compress appear three times in the output with the proposed implementation? If not, could the prefix be pulled up there..

If that isn't an option, I'd vote for changing the last two stats so they share a common prefix and will sort together in t2.

Comment (ajmorton) [reply dist/stat_data.py:834]: Yep it does. Example t2 statistic: `wt table compression compressed pages read`
I'll drop the first use of compression so stats will read as
`wt table compression pages read from disk with compression ratio smaller than  2`

Comment (ajmorton) [reply dist/stat_data.py]: Updated in [b575a62](https://github.com/wiredtiger/wiredtiger/pull/9908/commits/b575a62ad44bb6b3ca6992d30b7d0c6eef858b61)

**Review by etienneptl — APPROVED** [review]
LGTM!

**Review by agorrod — APPROVED** [review]
LGTM - thanks!

**Comment by ajmorton** [comment]
evergreen merge
